[Ring Series]: Add high-level Power Series Ring API with domain and element classes

Jatinbhardwaj-093 opened the pull request:
### Reference
Issue: #26957
Extend and wrap the lower class: #28109

### Brief Description

This implementation introduces a higher-level representation for the Power Series Ring, providing a structured way to represent both the ring itself and its elements, along with a proper domain.
 - Ring (PowerSeriesRingRing, PowerSeriesRIngField)
	- PowerSeriesRingRing - Provides core power series ring operations (add, multiply, invert, compose, differentiate, revert) over a base ring. Additionally, this class extends functionality with methods for converting or constructing ring elements from various types (e.g., expressions, lists, integers).
	- PowerSeriesRingField - Extends PowerSeriesRingRing for fields, adding functions like sqrt, integration, log, exp, and transcendental operations.
 - Element (PowerSeriesElement)
The PowerSeriesElement class provides a structured representation of ring elements, supporting standard arithmetic operations (+, -, /, …). This implementation also supports an explicit Order term. If the external order term exceeds the precision of the ring, it is truncated to the ring’s precision.
 - Domain (SeriesRing)
The Series domain class includes the associated ring (self.ring) as metadata, which can be extracted for ring-specific operations. It provides methods for converting elements from other domains into its own data type.

### Other Comments

The factory function `power_series_ring` has been updated to return a tuple of the PowerSeriesRing and its generator, rather than lower-level rings. This makes it similar to  `ring` function used for polynomials (**PolyRing**).

```python
>>> from sympy import QQ
>>> from sympy.polys.series import power_series_ring
>>> R, y = power_series_ring("y", QQ, 10)
>>> R.inverse(1 + y)
1 - y + y**2 - y**3 + y**4 - y**5 + y**6 - y**7 + y**8 - y**9 + O(y**10)
```

### Release Notes

<!-- BEGIN RELEASE NOTES -->
* polys
   * A new Power Series module (polys/series) introduces **PowerSeriesRing** for working with univariate power series truncated to finite precision. It provides two ring types: 
   		* **PowerSeriesRingRing** for ring domains such as ZZ, and 
   		* **PowerSeriesRingField** for field domains such as QQ. 	Current implementation for QQ supports Taylor expansions at `x = 0` for functions such as exp, log, sin, cos, and others with well-defined series of rational coefficients.
   * A helper function power_series_ring automatically selects the appropriate series ring based on the ground domain.
<!-- END RELEASE NOTES -->

Jatinbhardwaj-093 committed:
added three class domain(Series), ring(PowerSeriesRing), element(PowerSeriesElement).

Jatinbhardwaj-093 commented:
Not ready for any kind of review.

Jatinbhardwaj-093 committed:
test case for some method in ring

Jatinbhardwaj-093 committed:
added tests for rational value operations, add test for series expansion. Fixes minor bugs

Jatinbhardwaj-093 committed:
Udated and fix the docs

Jatinbhardwaj-093 committed:
remove feat: the addition of order term in series element.
white space fix in series.rst.
added right div.
added method order_term() which return order term of the ring prec.
some doc fix.

oscarbenjamin commented on sympy/polys/domains/powerseriesring.py:
```diff
@@ -0,0 +1,126 @@
+        """Convert SymPy's expression to `dtype`."""
+        return self.ring.from_expr(a)
+
+    def from_ZZ(K1, a, K0):
+        """Convert a Python `int` object to `dtype`."""
+        return K1(K1.domain.convert(a, K0))
```
I think it is generally better if these methods don't fallback on calling `convert` but rather call some specific function that handles the ZZ case. It might be necessary to add such methods to the ring.

Jatinbhardwaj-093 committed:
type fix

trailing space

Jatinbhardwaj-093 committed:
fix version issue

Jatinbhardwaj-093 committed:
doc fix

Jatinbhardwaj-093 committed:
added docs.
fix type issue.
add the method power_series_ring in domain class

.

oscarbenjamin commented on sympy/polys/domains/domain.py:
```diff
@@ -1012,6 +1013,11 @@ def poly_ring(self, *symbols: str | Expr, order: str | MonomialOrder = lex) -> P
         from sympy.polys.domains.polynomialring import PolynomialRing
         return PolynomialRing(self, symbols, order)
 
+    def power_series_ring(self, *symbols: str | Expr, prec: int = 6) -> PowerSeriesRing:
+        """Returns a Power Series ring over ring or field. """
+        from sympy.polys.domains.powerseriesring import PowerSeriesRing
+        return PowerSeriesRing(self, symbols[0], prec)
```
This should raise an error if multiple symbols are passed.

oscarbenjamin commented on sympy/polys/domains/powerseriesring.py:
```diff
@@ -0,0 +1,138 @@
+from sympy.polys.series.tring import TElement
+from sympy.utilities import public
+
+from typing import Union, TYPE_CHECKING
```
We don't need to use `Union` because it is just `a | b`.

oscarbenjamin commented on sympy/polys/domains/powerseriesring.py:
```diff
@@ -0,0 +1,138 @@
+    has_assoc_Ring = True
+    has_assoc_Field = False
+
+    ring: Union[PowerSeriesRingRing, PowerSeriesRingField]
```
This should be some kind of protocol like:
```
PowerSeriesRingProto[Er]
```
There should be a protocol that both classes match so that the type here is the protocol not a union. The type parameter Er needs to be here.

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1007 @@
+        self.symbol = symbol
+        self.prec = prec
+
+    @property
+    def ring(self) -> PowerSeriesRingProto[TElement, Er]:
+        """Return the base ring of the power series."""
+        return cast(
+            "PowerSeriesRingProto[TElement, Er]",
+            _power_series_ring(self.domain, self.prec),
+        )
```
The case should not be necessary either. The type annotations for the function being called should be fixed to return the correct type.

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1007 @@
+            _power_series_ring(self.domain, self.prec),
+        )
+
+    @property
+    def one(self) -> PowerSeriesElement[Er]:
+        return self.ground_new(self.domain.one)
+
+    @property
+    def zero(self) -> PowerSeriesElement[Er]:
+        return self.ground_new(self.domain.zero)
```
If the elements are immutable then this should be stored in the constructor as `self.one`.

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1007 @@
+        if fring_prec:
+            prec = self.prec
+
+        while coeffs and coeffs[-1] == self.domain.zero:
+            coeffs.pop()
```
I think it would be better to collect a dict for coeffs rather than making a (potentially large) list and popping off the zeros after.

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1007 @@
+        if not isinstance(other, PowerSeriesElement):
+            return False
+
+        R = self.ring.ring
```
I think that this is storing the wrong kind of ring as `self.ring`.

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1007 @@
+        series = R.add(self.series, other.series)
+        return self.new(series)
+
+    def _add_ground(self, other: Er) -> PowerSeriesElement[Er]:
+        s2 = self.ring.ground_new(other)
+        return self._add(s2)
```
Part of the point of methods like `add_ground` is not just to convert the type but to have a more efficient implementation. We should not need to create `s2` here.

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1007 @@
+
+        """
+        if not other:
+            return self
```
The type should be checked before doing this e.g. this is wrong:
```python
In [1]: x = QQ['x'].gens[0]

In [2]: x
Out[2]: x

In [3]: x + []
Out[3]: x
```
It should be a TypeError.

oscarbenjamin commented on sympy/polys/series/tring.py:
```diff
@@ -0,0 +1,89 @@
+        return cast("TSeriesRingField[MPQ]", R_flint)
+
+
+def _power_series_ring(K: Domain, prec: int = 6) -> TSeriesRing | TSeriesRingField:
+    """
+    Helper function for the Power Series Ring classes to create a base ring from lower
+    power series rings.
+    """
+    if K.is_ZZ:
+        return cast("TSeriesRing", PowerSeriesRingZZ(prec))
+    elif K.is_QQ:
+        return cast("TSeriesRingField", PowerSeriesRingQQ(prec))
+    else:
+        raise ValueError(f"Unsupported ground domain: {K}")
```
This function should be typed as
```python
@overload
def _power_series_ring(K: Field[Ef], prec: int = 6) -> TSeriesRingField[Ef]:
    ...
@overload
def _power_series_ring(K: Domain[Er], prec: int = 6) -> TSeriesRing[Er]:
    ...

def _power_series_ring(K: Domain[Er] | Field[Ef], prec: int = 6) -> TSeriesRing[Er] | TSeriesRingField[Ef]:
    """
    Helper function for the Power Series Ring classes to create a base ring from lower
    power series rings.
    """
    if K.is_ZZ:
        R_ZZ: TSeriesRing[MPZ] = PowerSeriesRingZZ(prec)
        return cast("TSeriesRing[Er]", R_ZZ)
    elif K.is_QQ:
        R_QQ: TSeriesRingField[MPQ] = PowerSeriesRingQQ(prec)
        return cast("TSeriesRingField[Ef]", R_QQ)
    else:
        raise ValueError(f"Unsupported ground domain: {K}")
```

oscarbenjamin commented on sympy/polys/series/tring.py:
```diff
@@ -0,0 +1,89 @@
+
+
+# Common type for the all lower power series elements.
+TElement = Union[USeries, ZZSeries, QQSeries]
```
I don't think that `TElement` should be used anywhere. Everywhere else should use `TSeriesElement[Er]` (or `TSeriesElement[Ef]`).

The purpose of the `_power_series_ring` function below is to make this seem like a concrete type everywhere else so that the actual underlying types like `ZZSeries` are obscured behind the opaque `TSeriesElement` type.

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1007 @@
         raise ValueError(f"Unsupported ground domain: {K}")
+
+
+class PowerSeriesRingRing(Generic[Er]):
+    """A class for representing Univariate Power Series Rings over a Ring."""
+
+    is_PowerSeries: bool = True
+    has_assoc_Ring: bool = True
+    has_assoc_Field: bool = False
```
These kinds of attributes are generally defined on Domain classes. I would not define them here.

Jatinbhardwaj-093 committed:
remove TElement, improve the types, fixes to improve add routine

reverting trail space

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1100 @@
+        1 + x - 1/2*x**2 - 1/6*x**3 + 1/24*x**4 + 1/120*x**5 + O(x**6)
+
+        """
+        if not other:
+            if isinstance(other, (PowerSeriesElement, int)) or self.domain.of_type(
+                other
+            ):
+                return self
+            raise TypeError(
+                f"Unsupported type {type(other).__name__} for addition with PowerSeriesElement."
+            )
```
This isn't the right way to handle this. It is better to remove the `if not other` check here at the start of the function. The types should be checked first (as they are below). Once we know we have the correct type we can check `if not other` if that is worth doing. Possibly the right place for this is in `_add` or `_add_ground` rather than here where we should really just be dispatching on types.

Jatinbhardwaj-093 committed:
added protocol for PowerSereisRing

Jatinbhardwaj-093 committed:
recursive from_expr

Jatinbhardwaj-093 committed:
remove unnecessary import, and if statement.

Jatinbhardwaj-093 committed:
full reference path for see also section.

Jatinbhardwaj-093 committed:
added is_ground, leading_coefficient to the base ring.

Jatinbhardwaj-093 committed:
Merge branch 'master' into series/domain-ring-element

Jatinbhardwaj-093 committed:
added domain in doc

.

oscarbenjamin commented on sympy/polys/series/base.py:
```diff
@@ -109,6 +109,18 @@ def equal_repr(self, s1: TSeries, s2: TSeries, /) -> bool:
+        ...
+
+    def leading_coefficient(self, s: TSeries) -> Er:
+        """Return the leading coefficient of a power series."""
```
I think it is better to call this the constant coefficient rather than the leading coefficient. For polynomials LC is the coefficient for the term of highest degree.

oscarbenjamin commented on sympy/polys/series/ringflint.py:
```diff
@@ -642,6 +680,26 @@ def equal_repr(self, s1: QQSeries, s2: QQSeries) -> bool:
+        """Check if a arg is an element of the power series ring."""
+        return isinstance(arg, (fmpq_poly, fmpq_series))
+
+    def is_ground(self, arg: QQSeries | MPQ) -> bool | None:
```
I think that the argument here should just be `QQSeries`. The purpose of this method is to check if a series is a ground element.

Is this method used anywhere though?

Jatinbhardwaj-093 replied:
it is used in the`is_unit` method of the PowerSeriesRing(Domain) class.

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1140 @@
+    def ring_new(self, arg) -> PowerSeriesElement[Er]: ...
+
+
+@overload
+def power_series_ring(
+    symbol: str, K: Domain[Ef], prec: int = 6
+) -> tuple[PowerSeriesRingField[Ef], PowerSeriesElement[Ef]]: ...
```
This should be `Field[Ef]`

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1140 @@
+        1 + x - 1/2*x**2 - 1/6*x**3 + 1/24*x**4 + 1/120*x**5 + O(x**6)
+
+        """
+        if not isinstance(other, (PowerSeriesElement, int)) and not self.domain.of_type(
+            other
+        ):
+            raise TypeError(
+                f"Unsupported type {type(other).__name__} for addition with PowerSeriesElement."
+            )
```
Methods like `__add__` should not raise `TypeError` or `NotImplementedError` for unrecognised types. They should return `NotImplemented`.

Only after confirming the type it can be possible to raise an error if the objects are incompatible e.g. for series with different symbols or different domains.

oscarbenjamin replied:
I explained this below but my comment was marked as resolved even though the code still does not return `NotImplemented`.

I don't know to what extent you are using AI helpers here but one thing I have noticed as that AI tools do not understand how to implement `__add__` correctly and how to use `NotImplemented` properly. I think this is because they have been trained on so much bad code and so many people get this wrong.

Jatinbhardwaj-093 committed:
method to do ground add, subtract and right subtract.
Made pow_int to handle -ve exponent if possible.
reduce the usage of to_list in element class

Jatinbhardwaj-093 committed:
improve test coverage

Jatinbhardwaj-093 committed:
fix the ring_new test

Jatinbhardwaj-093 committed:
use to_dense by ring as elemnent one is removed

oscarbenjamin commented on sympy/polys/series/ringflint.py:
```diff
@@ -642,6 +699,24 @@ def equal_repr(self, s1: QQSeries, s2: QQSeries) -> bool:
+        """Check if a arg is an element of the power series ring."""
+        return isinstance(arg, (fmpq_poly, fmpq_series))
+
+    def is_ground(self, arg: QQSeries) -> bool | None:
+        """Check if a arg is a ground element of the power series ring."""
+        if self.prec == 0:
+            return None
+
+        if self.is_element(arg):
+            return len(arg) <= 1
+        else:
+            return False
```
This method seems a bit confused. The argument is annotated as `QQSeries` but then it checks if the arg is an element and otherwise returns False.

I think this should either assume that the argument is an element or if it is going to check then it should otherwise raise TypeError.

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1081 @@
 
-    FlintPowerSeriesRingZZ: type[_FlintPowerSeriesRingZZ] | None
-    FlintPowerSeriesRingQQ: type[_FlintPowerSeriesRingQQ] | None
+T = TypeVar("T")
```
It doesn't look like this type variable is used anywhere.

oscarbenjamin commented on sympy/polys/domains/domain.py:
```diff
@@ -1054,6 +1055,13 @@ def poly_ring(self, *symbols: str | Expr, order: str | MonomialOrder = lex) -> P
         return PolynomialRing(self, symbols, order)
 
+    def power_series_ring(self, *symbols: str | Expr, prec: int = 6) -> PowerSeriesRing:
+        """Returns a Power Series ring over ring or field. """
```
This should have a better docstring including links to the polys/series section of the docs.

oscarbenjamin replied:
Let's make this function private for now (with a leading underscore).

Also the docs for the power series ring domain should note that it is experiemental.

A followup PR can be made for this that tests that all the pieces are working with DomainMatrix and Poly.

oscarbenjamin commented on sympy/polys/domains/powerseriesring.py:
```diff
@@ -0,0 +1,139 @@
+        """Check if ``a`` is of type ``dtype``."""
+        return self.ring.is_element(element)
+
+    @property
+    def zero(self) -> PowerSeriesElement[Er]:  # type: ignore
+        return self.ring.zero
+
+    @property
+    def one(self) -> PowerSeriesElement[Er]:  # type: ignore
+        return self.ring.one
```
This should use `@cached_property`.

oscarbenjamin commented on sympy/polys/domains/powerseriesring.py:
```diff
@@ -0,0 +1,139 @@
+        """Convert SymPy's expression to `dtype`."""
+        return self.ring.from_expr(a)
+
+    def from_ZZ(K1, a, K0):
+        """Convert a Python `int` object to `dtype`."""
+        return K1(K1.domain.convert(a, K0))
```
It should be `K1.ring.ground_new(K1.convert(...))`.

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1081 @@
+        return self.new(s)
+
+    def new(self, series: TSeriesElement[Er]) -> PowerSeriesElement[Er]:
+        return self.__class__(self._parent_ring, series)
```
I would make this private (leading underscore)

Jatinbhardwaj-093 committed:
change the method ground_new -> from_ground, add from_int both are srictly typed.
remove unnecessary ovehead symbols, domain parameter from the element class attribute.
move SeriesRingProto to the powerseriesring file.
remove is_element from lower rings.

quality fix

Jatinbhardwaj-093 committed:
fix doc : SeriesRingProto

oscarbenjamin commented on doc/src/modules/polys/series.rst:
```diff
@@ -25,56 +25,92 @@ The architecture includes two backends: a pure Python implementation and a
     14*x - 9*x**2 - 16*x**3 - 3*x**4
 
 However, if an operation produces a result that exceeds the precision threshold,
 it is automatically truncated.
```
I think it would be good to demonstrate below that a series can be constructed with an explicit precision that is either above or below the ring precision:
```python
In [14]: R, x = power_series_ring('x', ZZ, 3)

In [15]: R.from_list([1, 1], prec=2)
Out[15]: 1 + x + O(x**2)

In [16]: R.from_list([1, 1, 3, 4], prec=4)
Out[16]: 1 + x + 3*x**2 + O(x**3)
```
Wait should that last one return with `O(x**4)`?

oscarbenjamin commented on sympy/polys/series/tring.py:
```diff
@@ -0,0 +1,99 @@
+    flint = False
+
+
+class TSeriesElement(Generic[Er]):
+    """Dummy type for lower power series elements, used only for static type checking."""
+
+    # Dummy method to ensure that TSeriesElement is invariant in Er.
+    def __invalid__(self, other: Er) -> Er:
+        raise NotImplementedError
+
+    flint = True
```
Why does this have `flint = True`?

oscarbenjamin commented:
Can you update the release notes in the OP?

I'm not sure what release notes are currently there for this project but there should be a bigger release note that explains the bigger picture of the new `polys.series` module.

Jatinbhardwaj-093 committed:
flint _get_series_prec if the version above 0.8 it can directly get the series prec by prec attribure

Jatinbhardwaj-093 committed:
rebase master

oscarbenjamin commented on sympy/polys/series/ring.py:
```diff
@@ -1,109 +1,1048 @@
+        elif self.domain.of_type(arg):
+            return self.from_ground(arg)
+        else:
+            return NotImplemented
```
This should be `raise TypeError`.

`NotImplemented` should only be returned from binary dunder methods like `__add__`.

Jatinbhardwaj-093 committed:
fix hypthesis test

oscarbenjamin commented:
Okay, looks good. Let's get this in.
